crates/factor-blobstore/src/host.rs

```diff
@@ -0,0 +1,536 @@
+    async fn delete_objects(&self, names: &[String]) -> anyhow::Result<()>;
+    async fn has_object(&self, name: &str) -> anyhow::Result<bool>;
+    async fn object_info(&self, name: &str) -> anyhow::Result<bs::types::ObjectMetadata>;
+    async fn get_data(
+        &self,
+        name: &str,
+        start: u64,
+        end: u64,
+    ) -> anyhow::Result<Box<dyn IncomingData>>;
+    async fn connect_stm(
+        &self,
+        name: &str,
+        stm: ReadHalf<SimplexStream>,
+        finished_tx: mpsc::Sender<anyhow::Result<()>>,
+    ) -> anyhow::Result<()>;
+    async fn list_objects(&self) -> anyhow::Result<Box<dyn ObjectNames>>;
+}
+
+#[async_trait]
+pub trait ObjectNames: Send + Sync {
+    async fn read(&mut self, len: u64) -> anyhow::Result<(Vec<String>, bool)>;
+    async fn skip(&mut self, num: u64) -> anyhow::Result<(u64, bool)>;
+}
+
+#[async_trait]
+pub trait IncomingData: Send + Sync {
+    async fn consume_sync(&mut self) -> anyhow::Result<Vec<u8>>;
+    fn consume_async(&mut self) -> wasmtime_wasi::pipe::AsyncReadStream;
+    async fn size(&mut self) -> anyhow::Result<u64>;
+}
+
+pub struct OutgoingValue {
+    read: Option<ReadHalf<SimplexStream>>,
+    write: Option<WriteHalf<SimplexStream>>,
+    stop_tx: Option<mpsc::Sender<()>>,
+    finished_rx: Option<mpsc::Receiver<anyhow::Result<()>>>,
+}
+
+const OUTGOING_VALUE_BUF_SIZE: usize = 16 * 1024;
+
+impl OutgoingValue {
+    fn new() -> Self {
+        let (read, write) = tokio::io::simplex(OUTGOING_VALUE_BUF_SIZE);
+        Self {
+            read: Some(read),
+            write: Some(write),
+            stop_tx: None,
+            finished_rx: None,
+        }
+    }
+
+    fn write_stream(&mut self) -> anyhow::Result<crate::AsyncWriteStream> {
+        let Some(write) = self.write.take() else {
+            anyhow::bail!("OutgoingValue has already returned its write stream");
+        };
+
+        let (stop_tx, stop_rx) = mpsc::channel(1);
+
+        self.stop_tx = Some(stop_tx);
+
+        let stm = crate::AsyncWriteStream::new_closeable(OUTGOING_VALUE_BUF_SIZE, write, stop_rx);
+        Ok(stm)
+    }
+
+    fn syncers(
+        &mut self,
+    ) -> (
+        Option<&mpsc::Sender<()>>,
+        Option<&mut mpsc::Receiver<anyhow::Result<()>>>,
+    ) {
+        (self.stop_tx.as_ref(), self.finished_rx.as_mut())
+    }
+
+    fn take_read_stream(
+        &mut self,
+    ) -> anyhow::Result<(ReadHalf<SimplexStream>, mpsc::Sender<anyhow::Result<()>>)> {
+        let Some(read) = self.read.take() else {
+            anyhow::bail!("OutgoingValue has already been connected to a blob");
+        };
+
+        let (finished_tx, finished_rx) = mpsc::channel(1);
+        self.finished_rx = Some(finished_rx);
+
+        Ok((read, finished_tx))
+    }
+}
+
+#[async_trait]
+pub trait Finishable: Send + Sync {
+    async fn finish(&mut self);
+}
+
+pub struct BlobStoreDispatch<'a> {
+    allowed_containers: HashSet<String>,
```

Comment:
I'm not _entirely_ sure that it can be but if this can be borrowed it would save a clone
```suggestion
    allowed_containers: &'a HashSet<String>,
```

Reply:
Actually if this one works then probably all of the `Arc`s below could just be borrows. You'd still need the locks for interior mutability of course.

Reply:
@lann I tried it and it hasn't segfaulted yet - thanks for the suggestion!